Add custom exception handlers to generated COM wrappers

dongle-the-gadget opened the pull request:
Fixes #109522

dongle-the-gadget committed:
Add custom exception handler to COM Wrappers

This is an initial iteration to add custom exception handlers to
generated COM wrappers, through a new ExceptionToUnmanagedMarshaller
property in GeneratedComWrappersAttribute

Fix #109522

dongle-the-gadget commented:
@dotnet-policy-service agree

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs:
```diff
@@ -396,7 +396,7 @@ private static IncrementalMethodStubGenerationContext CalculateStubInformation(M
                 callConv.ToSequenceEqualImmutableArray(SyntaxEquivalentComparer.Instance),
                 virtualMethodIndexData,
-                new ComExceptionMarshalling(),
+                new ComExceptionMarshalling(exceptionToUnmanagedMarshallerQualifiedName),
```
Instead of editing `ComExceptionMarshalling`, please use the same logic as `ExceptionMarshalling.Custom`.

`ComExceptionMarshalling` specifically represents the default behavior for COM. Custom exception marshallers should replace that behavior with a different marshaller, not try to hook into it.

dongle-the-gadget replied:
Finished. I've kept the accessible-by-file-modifier logic in here as I'm not sure if that is checked by `CustomMarshallingInfoHelper.CreateNativeMarshallingInfoForNonSignatureElement` or not. If it is then we could cut down on the added resources as well.

jkoritzinsky replied:
I don't believe that's already checked there, so lets keep the diagnostic.

dongle-the-gadget committed:
Copy VTable Index exception logic over

Resolves an issue with previous implementation, highlighted by Jeremy.
The previous implementation is overcomplicated and hooks into the
default exception logic, which is not ideal. This commit copies the
existing custom exception marshalling logic from the VTable stub
generator.

dongle-the-gadget committed:
Remove redundant marshaller in interface data

dongle-the-gadget committed:
Remove redundant metadata name logic

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/ISymbolExtensions.cs:
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Diagnostics.CodeAnalysis;
+using System.Text;
```
```suggestion
```

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/BoundGenerators.cs:
```diff
@@ -84,7 +84,7 @@ public static BoundGenerators Create(ImmutableArray<TypePositionInfo> elementTyp
                 }
 
-                if (managedExceptionInfo.MarshallingAttributeInfo is ComExceptionMarshalling)
+                if (managedExceptionInfo.MarshallingAttributeInfo is ComExceptionMarshalling comExceptionMarshalling)
```
```suggestion
                if (managedExceptionInfo.MarshallingAttributeInfo is ComExceptionMarshalling)
```

jkoritzinsky reviewed (commented):
Just a few nits, but looking much better!

dongle-the-gadget committed:
Remove leftover System.Text namespace

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>

dongle-the-gadget committed:
Remove leftover null pattern matching

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/GeneratedComInterfaceAttributeData.cs:
```diff
@@ -24,7 +24,7 @@ public static GeneratedComInterfaceData From(GeneratedComInterfaceCompilationDat
                     ? ManagedTypeInfo.CreateTypeInfoForTypeSymbol(generatedComInterfaceAttr.StringMarshallingCustomType)
                     : null,
-                Options = generatedComInterfaceAttr.Options
+                Options = generatedComInterfaceAttr.Options,
```
```suggestion
                Options = generatedComInterfaceAttr.Options
```

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/GeneratedComInterfaceAttributeData.cs:
```diff
@@ -70,6 +71,17 @@ public static GeneratedComInterfaceCompilationData GetDataFromAttribute(Attribut
+            {
+                if (exceptionToUnmanagedMarshaller.Value is not INamedTypeSymbol)
+                {
+                    return null;
```
We generally don't use `null` for logical flow. This would mean the return type should be updated. I don't think we should be failing in this case as it will silent and difficult to reason about. A diagnostic would be more appropriate.

dongle-the-gadget replied:
This was based on [`InteropAttributeDataExtensions.WithValuesFromNamedArguments`](https://github.com/dotnet/runtime/blob/07f3e33da71e52fce0627aa8645fc9787b954672/src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/InteropAttributeData.cs#L82-L90), which returns `null` if an invalid `StringMarshalling` type is passed. I'm not sure how this is handled particularly by `GeneratedComInterfaceAttributeData`.

jkoritzinsky replied:
I think this is another one of the cases where we use `null` as our "unable to parse" sentinel (and in the caller we check for null and emit a diagnostic and replace with a sentinel). Can you confirm?

dongle-the-gadget replied:
Unfortunately, it looks like there's no regards to null-checking here. Even today if you pass in an invalid type for `StringMarshallingCustomType`, i.e.
```cs
[GeneratedComInterface(StringMarshalling = StringMarshalling.Custom, StringMarshallingCustomType = typeof(string*))]
public partial interface Foo;
```
the generator fails with a generic error message from Roslyn: 
> Generator 'ComInterfaceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'NullReferenceException' with message 'Object reference not set to an instance of an object.'

dongle-the-gadget replied:
This is done.

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/GeneratedComInterfaceAttributeData.cs:
```diff
@@ -35,6 +35,7 @@ public static GeneratedComInterfaceData From(GeneratedComInterfaceCompilationDat
     internal sealed record GeneratedComInterfaceCompilationData : InteropAttributeCompilationData
     {
         public ComInterfaceOptions Options { get; init; } = ComInterfaceOptions.ManagedObjectWrapper | ComInterfaceOptions.ComObjectWrapper;
+        public INamedTypeSymbol? ExceptionToUnmanagedMarshaller { get; init; }
```
I don't think this should ever be nullable. If the user didn't supply it we should be able to contruct the default one instead and avoid using a `INamedTypeSymbol`.

jkoritzinsky replied:
I disagree. Coming from the attribute data itself, there's no good named type symbol to put here for the default case. What we use depends on the return type, and `null` is the value of the field in the attribute at runtime (if someone were to call GetAttribute).

AaronRobinsonMSFT replied:
I see. This looks like an implementation detail I'm missing. If `null` makes sense I'd hope we don't see `null` after `GeneratedComInterfaceCompilationData` is processed.

jkoritzinsky replied:
This is converted from a Roslyn symbol to our data model in `CalculateStubInformation` so from that point onwards there's no `null`.

dongle-the-gadget committed:
Remove extra colon

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

dongle-the-gadget committed:
Prevent exception when exception marshaller is invalid

dongle-the-gadget committed:
Merge branch 'issue-109522' of https://github.com/dongle-the-gadget/runtime into issue-109522

jkoritzinsky reviewed (commented):
Changes look good. Mark this as non-draft and we can do a final review. (Also get the new test passing everywhere, looks like there's failures).

dongle-the-gadget committed:
Fix exception marshalling test.

jkoritzinsky reviewed (approved):
LGTM Thank you!

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs:
```diff
@@ -389,14 +389,30 @@ private static IncrementalMethodStubGenerationContext CalculateStubInformation(M
 
+            MarshallingInfo exceptionMarshallingInfo;
+
+            if (generatedComInterfaceAttributeData.ExceptionToUnmanagedMarshaller is not null)
+            {
+                exceptionMarshallingInfo = CustomMarshallingInfoHelper.CreateNativeMarshallingInfoForNonSignatureElement(
+                    environment.Compilation.GetTypeByMetadataName(TypeNames.System_Exception),
+                    (INamedTypeSymbol)generatedComInterfaceAttributeData.ExceptionToUnmanagedMarshaller,
+                    generatedComAttribute,
+                    environment.Compilation,
+                    generatorDiagnostics);
+            }
+            else
+            {
+                exceptionMarshallingInfo = new ComExceptionMarshalling();
+            }
```
```suggestion
            if (generatedComInterfaceAttributeData.ExceptionToUnmanagedMarshaller is null)
            {
                exceptionMarshallingInfo = new ComExceptionMarshalling();
            }
            else
            {
                exceptionMarshallingInfo = CustomMarshallingInfoHelper.CreateNativeMarshallingInfoForNonSignatureElement(
                    environment.Compilation.GetTypeByMetadataName(TypeNames.System_Exception),
                    (INamedTypeSymbol)generatedComInterfaceAttributeData.ExceptionToUnmanagedMarshaller,
                    generatedComAttribute,
                    environment.Compilation,
                    generatorDiagnostics);
            }
```

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/CompileFails.cs:
I don't see a passing test for consuming a type. We should add that as well.

dongle-the-gadget replied:
I've put together an end-to-end test, although the Compiles unit test is rather complex for me.

dongle-the-gadget committed:
Flip exception marshaller cases

This is to match the common scenario.

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

dongle-the-gadget committed:
Add custom exception marshaller test.

dongle-the-gadget committed:
Update exception marshaller to check for return value

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Tests/ExceptionToUnmanagedMarshallerTests.cs:
```diff
@@ -0,0 +1,54 @@
+            CustomExceptionMarshallerComClass comObject = new();
+            StrategyBasedComWrappers wrappers = new();
+            nint nativeUnknown = wrappers.GetOrCreateComInterfaceForObject(comObject, CreateComInterfaceFlags.None);
+            Marshal.QueryInterface(nativeUnknown, typeof(ICustomExceptionMarshallerComInterface).GUID, out nint nativeInterface);
```
This isn't unreasonable but not our normal process here. What I would do is create a wrapping RCW for the CCW, just call `GetOrCreateObjectForComInterface()`, don't unwrap it. This will create the following structure we can test:

`CustomExceptionMarshallerComClass` -> `CCW` -> `RCW` -> `ICustomExceptionMarshallerComInterface`

At this point you can use the `ICustomExceptionMarshallerComInterface` and catch the expected exception and will verify the full stack. It will also handle the potential leaking that is occuring for the two `Marshal.Release()` calls and exposing the details of the vtable layout.

dongle-the-gadget committed:
Create RCW in exception marshaller test.

AaronRobinsonMSFT committed:
Merge branch 'main' into issue-109522
